
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Python library for genome annotations">
      
      
      
        <link rel="canonical" href="https://pcingola.github.io/pygnome/format_specifications/vcf/">
      
      
      
      
      <link rel="icon" href="../../pygnome.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Vcf - PyGnome</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#1-the-vcf-specification" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="PyGnome" class="md-header__button md-logo" aria-label="PyGnome" data-md-component="logo">
      
  <img src="../../pygnome.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PyGnome
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Vcf
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/pcingola/pygnome" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    pcingola/pygnome
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="PyGnome" class="md-nav__button md-logo" aria-label="PyGnome" data-md-component="logo">
      
  <img src="../../pygnome.png" alt="logo">

    </a>
    PyGnome
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pcingola/pygnome" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    pcingola/pygnome
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../getting-started.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    User Guide
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/installation.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/basic-usage.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic Usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/advanced-usage.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Advanced Usage
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Format Specifications
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Format Specifications
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../format-specifications/fasta.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    FASTA
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../format-specifications/fastq.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    FASTQ
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../format-specifications/gff3.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    GFF3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../format-specifications/gtf.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    GTF
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../format-specifications/vcf.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    VCF
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/index.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/genomics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Genomics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/parsers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Parsers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/sequences/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/feature-store.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feature Store
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Development
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../development/contributing.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../development/changelog.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-the-vcf-specification" class="md-nav__link">
    <span class="md-ellipsis">
      1 The VCF specification
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1 The VCF specification">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-an-example" class="md-nav__link">
    <span class="md-ellipsis">
      1.1 An example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-character-encoding-non-printable-characters-and-special-characters" class="md-nav__link">
    <span class="md-ellipsis">
      1.2 Character encoding, non-printable characters and special characters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-data-types" class="md-nav__link">
    <span class="md-ellipsis">
      1.3 Data types
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-metainformation-lines" class="md-nav__link">
    <span class="md-ellipsis">
      1.4 Meta‑information lines
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.4 Meta‑information lines">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#141-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.1 File format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#143-filter-field-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.3 Filter field format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#145-alternative-allele-field-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.5 Alternative allele field format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#146-assembly-field-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.6 Assembly field format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#148-sample-field-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.8 Sample field format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#149-pedigree-field-format" class="md-nav__link">
    <span class="md-ellipsis">
      1.4.9 Pedigree field format
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#15-header-line-syntax" class="md-nav__link">
    <span class="md-ellipsis">
      1.5 Header line syntax
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#16-data-lines" class="md-nav__link">
    <span class="md-ellipsis">
      1.6 Data lines
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.6 Data lines">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#161-fixed-fields" class="md-nav__link">
    <span class="md-ellipsis">
      1.6.1 Fixed fields
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#162-genotype-fields" class="md-nav__link">
    <span class="md-ellipsis">
      1.6.2 Genotype fields
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1.6.2 Genotype fields">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#table-1-reserved-info-keys-excerpt" class="md-nav__link">
    <span class="md-ellipsis">
      Table 1: Reserved INFO keys (excerpt)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#table-2-reserved-formatgenotype-keys-excerpt" class="md-nav__link">
    <span class="md-ellipsis">
      Table 2: Reserved FORMAT/Genotype keys (excerpt)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-understanding-the-vcf-format-and-the-haplotype-representation" class="md-nav__link">
    <span class="md-ellipsis">
      2 Understanding the VCF format and the haplotype representation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2 Understanding the VCF format and the haplotype representation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-vcf-tag-naming-conventions" class="md-nav__link">
    <span class="md-ellipsis">
      2.1 VCF tag naming conventions
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-info-keys-used-for-structural-variants" class="md-nav__link">
    <span class="md-ellipsis">
      3 INFO keys used for structural variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-format-keys-used-for-structural-variants" class="md-nav__link">
    <span class="md-ellipsis">
      4 FORMAT keys used for structural variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-representing-variation-in-vcf-records" class="md-nav__link">
    <span class="md-ellipsis">
      5 Representing variation in VCF records
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5 Representing variation in VCF records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#51-creating-vcf-entries-for-snps-and-small-indels" class="md-nav__link">
    <span class="md-ellipsis">
      5.1 Creating VCF entries for SNPs and small indels
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5.1 Creating VCF entries for SNPs and small indels">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#511-example-1" class="md-nav__link">
    <span class="md-ellipsis">
      5.1.1 Example 1
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#512-example-2" class="md-nav__link">
    <span class="md-ellipsis">
      5.1.2 Example 2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#513-example-3" class="md-nav__link">
    <span class="md-ellipsis">
      5.1.3 Example 3
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#52-decoding-vcf-entries-for-snps-and-small-indels" class="md-nav__link">
    <span class="md-ellipsis">
      5.2 Decoding VCF entries for SNPs and small indels
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#53-encoding-structural-variants" class="md-nav__link">
    <span class="md-ellipsis">
      5.3 Encoding Structural Variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#54-specifying-complex-rearrangements-with-breakends" class="md-nav__link">
    <span class="md-ellipsis">
      5.4 Specifying complex rearrangements with breakends
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Vcf</h1>

<p>Below is the content of pages 4–16 of the VCF 4.5 PDF, faithfully converted into Markdown. Section and subsection numbers match the PDF.</p>
<h2 id="1-the-vcf-specification">1 The VCF specification</h2>
<p>Reference: https://samtools.github.io/hts-specs/VCFv4.5.pdf</p>
<h3 id="11-an-example">1.1 An example</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>##fileformat=VCFv4.5
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>##fileDate=20090805
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>##source=myImputationProgramV3.1
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>##contig=&lt;ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species=&quot;Homo sapiens&quot;,taxonomy=x&gt;
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>##phasing=partial
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>##INFO=&lt;ID=NS,Number=1,Type=Integer,Description=&quot;Number of Samples With Data&quot;&gt;
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a>##INFO=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Total Depth&quot;&gt;
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a>##INFO=&lt;ID=AF,Number=A,Type=Float,Description=&quot;Allele Frequency&quot;&gt;
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a>##INFO=&lt;ID=AA,Number=1,Type=String,Description=&quot;Ancestral Allele&quot;&gt;
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>##INFO=&lt;ID=DB,Number=0,Type=Flag,Description=&quot;dbSNP membership, build 129&quot;&gt;
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>##INFO=&lt;ID=H2,Number=0,Type=Flag,Description=&quot;HapMap2 membership&quot;&gt;
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>##FILTER=&lt;ID=q10,Description=&quot;Quality below 10&quot;&gt;
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>##FILTER=&lt;ID=s50,Description=&quot;Less than 50% of samples have data&quot;&gt;
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description=&quot;Genotype Quality&quot;&gt;
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Read Depth&quot;&gt;
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a>##FORMAT=&lt;ID=HQ,Number=2,Type=Integer,Description=&quot;Haplotype Quality&quot;&gt;
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a>#CHROM POS      ID          REF ALT    QUAL FILTER INFO                       FORMAT        NA00001         NA00002        NA00003
<a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a>20     14370    rs6054257   G   A      29   PASS   NS=3;DP=14;AF=0.5;DB;H2     GT:GQ:DP:HQ   0|0:48:1:51,51   1|0:48:8:51,51  1/1:43:5:.,.  
<a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a>20     17330    .           T   A      3    q10    NS=3;DP=11;AF=0.017           GT:GQ:DP:HQ   0|0:49:3:58,50   0|1:3:5:65,3    0/0:41:3     
<a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a>20     1110696 rs6040355   A   G,T    67   PASS   NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ 1|2:21:6:23,27   2|1:2:0:18,2    2/2:35:4     
<a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a>20     1230237 .           T   .      47   PASS   NS=3;DP=13;AA=T               GT:GQ:DP:HQ   0|0:54:7:56,60   0|0:48:4:51,51  0/0:61:2    
<a id="__codelineno-0-24" name="__codelineno-0-24" href="#__codelineno-0-24"></a>20     1234567 microsat1    GTC G,GTCT 50   PASS   NS=3;DP=9;AA=G              GT:GQ:DP      0/1:35:4       0/2:17:2       1/1:40:3    
</code></pre></div>
<p>This shows a simple SNP, a filtered‑out SNP, a multi‑allelic site, a monomorphic reference site, and a microsatellite with two alternate alleles; genotype columns illustrate phased vs. unphased calls and per‑sample quality/depth/HQ .</p>
<hr />
<h3 id="12-character-encoding-non-printable-characters-and-special-characters">1.2 Character encoding, non-printable characters and special characters</h3>
<p>VCF files must be UTF‑8 encoded (no byte‑order mark) and may use either LF (<code>\n</code>) or CR+LF (<code>\r\n</code>) as line separators. Non‑printable characters U+0000–U+0008, U+000B–U+000C, and U+000E–U+001F are disallowed. Characters with special meaning (e.g. <code>;</code>, <code>:</code>, <code>=</code>, <code>%</code>, <code>,</code>) within fields must be percent‑encoded as follows:</p>
<table>
<thead>
<tr>
<th>Encoding</th>
<th>Character</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>%3A</code></td>
<td><code>:</code></td>
<td></td>
</tr>
<tr>
<td><code>%3B</code></td>
<td><code>;</code></td>
<td></td>
</tr>
<tr>
<td><code>%3D</code></td>
<td><code>=</code></td>
<td></td>
</tr>
<tr>
<td><code>%25</code></td>
<td><code>%</code></td>
<td></td>
</tr>
<tr>
<td><code>%2C</code></td>
<td><code>,</code></td>
<td></td>
</tr>
<tr>
<td><code>%0D</code></td>
<td>CR</td>
<td></td>
</tr>
<tr>
<td><code>%0A</code></td>
<td>LF</td>
<td></td>
</tr>
<tr>
<td><code>%09</code></td>
<td>TAB</td>
<td></td>
</tr>
</tbody>
</table>
<hr />
<h3 id="13-data-types">1.3 Data types</h3>
<p>Supported VCF data types (applies to both INFO and FORMAT fields) are:</p>
<ul>
<li><strong>Integer</strong> (32‑bit signed; values ≤ ±2³¹–1)</li>
<li><strong>Float</strong> (32‑bit IEEE‑754; regex <code>^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$</code> or <code>^[-+]?(INF|INFINITY|NAN)$</code>)</li>
<li><strong>Flag</strong> (present/absent, Number=0)</li>
<li><strong>Character</strong> (single‐character string)</li>
<li><strong>String</strong> (arbitrary string)&#x20;</li>
</ul>
<hr />
<h3 id="14-metainformation-lines">1.4 Meta‑information lines</h3>
<p>Meta‑information lines begin with <code>##</code> and must come before the single header line (<code>#CHROM ...</code>). They are either:</p>
<ul>
<li><strong>Unstructured</strong>: <code>##key=value</code></li>
<li><strong>Structured</strong>: <code>##key=&lt;key1=val1,key2=val2,...&gt;</code>
  – All structured lines require a unique <code>ID=</code> field inside the <code>&lt;...&gt;</code> .</li>
</ul>
<h4 id="141-file-format">1.4.1 File format</h4>
<p>A required first line specifying the VCF version:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>##fileformat=VCFv4.5
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>``` :contentReference[oaicite:4]{index=4}
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>#### 1.4.2 Information field format
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>Defines INFO keys:
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>```vcf
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>##INFO=&lt;ID=ID,Number=&lt;number&gt;,Type=&lt;type&gt;,Description=&quot;...&quot;,Source=&quot;...&quot;,Version=&quot;...&quot;&gt;
</code></pre></div>
<ul>
<li><code>Type</code> ∈ {Integer, Float, Flag, Character, String}</li>
<li>
<p><code>Number</code> ∈ ℕ or:</p>
</li>
<li>
<p><code>A</code>: one value per ALT allele</p>
</li>
<li><code>R</code>: one value per allele (REF + ALTs)</li>
<li><code>G</code>: one value per genotype (see 1.6.2)</li>
<li><code>.</code>: unknown/unbounded</li>
<li><code>Flag</code> fields have <code>Number=0</code> and no <code>=value</code> part .</li>
</ul>
<h4 id="143-filter-field-format">1.4.3 Filter field format</h4>
<p>Defines FILTER codes:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>##FILTER=&lt;ID=ID,Description=&quot;...&quot;&gt;
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>``` :contentReference[oaicite:6]{index=6}
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>#### 1.4.4 Individual format field format
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>Defines per‑sample FORMAT keys:
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>```vcf
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>##FORMAT=&lt;ID=ID,Number=&lt;number&gt;,Type=&lt;type&gt;,Description=&quot;...&quot;&gt;
</code></pre></div>
<ul>
<li>Same types as INFO</li>
<li>
<p>Additional <code>Number</code> codes:</p>
</li>
<li>
<p><code>LA</code>, <code>LR</code>, <code>LG</code>: allele‐subset variants</p>
</li>
<li><code>P</code>: one value per allele in GT</li>
<li><code>M</code>: one value per possible base modification (see spec) .</li>
</ul>
<h4 id="145-alternative-allele-field-format">1.4.5 Alternative allele field format</h4>
<p>Defines symbolic ALT alleles (e.g., structural variants or ambiguity codes):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>##ALT=&lt;ID=type,Description=&quot;...&quot;&gt;
</code></pre></div>
<ul>
<li>Recommended types: <code>DEL</code>, <code>INS</code>, <code>DUP</code>, <code>INV</code>, <code>CNV</code>, with subtypes like <code>CNV:TR</code>, <code>DUP:TANDEM</code>, etc.</li>
<li>IUPAC codes as ALT can be declared (e.g. <code>ID=R</code> for A/G) .</li>
</ul>
<h4 id="146-assembly-field-format">1.4.6 Assembly field format</h4>
<p>(Optional) link to FASTA for breakend assemblies:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>##assembly=url
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>``` :contentReference[oaicite:9]{index=9}
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>#### 1.4.7 Contig field format
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>Defines contigs and optional metadata:
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>```vcf
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>##contig=&lt;ID=ctg1,length=...,md5=...,URL=...&gt;
</code></pre></div>
<ul>
<li><code>ID</code>: name matching <code>[!-~]</code> except certain punctuation</li>
<li>Optional <code>length</code>, <code>md5</code>, <code>URL</code> fields .</li>
</ul>
<h4 id="148-sample-field-format">1.4.8 Sample field format</h4>
<p>Defines sample metadata:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>##SAMPLE=&lt;ID=Sample1,Assay=WholeGenome,Ethnicity=AFR,...,Description=&quot;...&quot;&gt;
</code></pre></div>
 :contentReference[oaicite:11]{index=11}</p>
<h4 id="149-pedigree-field-format">1.4.9 Pedigree field format</h4>
<p>Defines pedigree relationships:
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>##PEDIGREE=&lt;ID=ChildID,Father=FatherID,Mother=MotherID&gt;
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>##pedigreeDB=url
</code></pre></div>
:contentReference[oaicite:12]{index=12}</p>
<hr />
<h3 id="15-header-line-syntax">1.5 Header line syntax</h3>
<p>The single header line names the fixed columns (tab‑delimited, no trailing tab):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>#CHROM POS ID REF ALT QUAL FILTER INFO [FORMAT [&lt;sampleID&gt;...]]
</code></pre></div>
<hr />
<h3 id="16-data-lines">1.6 Data lines</h3>
<p>All data lines are tab‑delimited, no trailing tab, and end with a newline. Missing values are <code>.</code>.</p>
<h4 id="161-fixed-fields">1.6.1 Fixed fields</h4>
<p>Each record has <strong>8</strong> mandatory fields:</p>
<ol>
<li><strong>CHROM</strong> (String): reference contig or <code>&lt;ID&gt;</code>; entries for one CHROM must be contiguous.  </li>
<li><strong>POS</strong> (Integer): 1‑based reference position; telomeres at 0 or N+1.  </li>
<li><strong>ID</strong> (String): semicolon‑sep. unique IDs or <code>.</code>.  </li>
<li><strong>REF</strong> (String): one or more bases (<code>A,C,G,T,N</code>), padded for indels at start; case-insensitive.  </li>
<li><strong>ALT</strong> (String): comma‑sep. alternate alleles (<code>A,C,G,T,N</code>, <code>*</code>, <code>.</code>, <code>&lt;ID&gt;</code>, <code>&lt;*&gt;</code>, breakends).  </li>
<li><strong>QUAL</strong> (Float): phred-scaled quality (<code>–10 log₁₀</code> error prob.), or <code>.</code>.  </li>
<li><strong>FILTER</strong> (String): <code>PASS</code> or semicolon‑sep. filter IDs; <code>.</code> if not applied.  </li>
<li><strong>INFO</strong> (String): semicolon‑sep. <code>key[=value[,...]]</code> entries or <code>.</code>; values may use percent‐encoding.  </li>
</ol>
<p>Reserved INFO keys are detailed in Table 1 below. :contentReference[oaicite:14]{index=14}</p>
<hr />
<h4 id="162-genotype-fields">1.6.2 Genotype fields</h4>
<p>If genotypes are present, a <code>FORMAT</code> column lists per‑sample keys (colon‑delimited), followed by one column per sample.  </p>
<ul>
<li>The first key <strong>must</strong> be <code>GT</code> if present; local‑allele fields require <code>LAA</code>.  </li>
<li>Missing values use <code>.</code>; trailing fields may be omitted except <code>GT</code>.  </li>
</ul>
<p>Common reserved FORMAT keys (Table 2) include <code>AD</code>, <code>DP</code>, <code>FT</code>, <code>GL</code>, <code>GP</code>, <code>GQ</code>, <code>GT</code>, <code>HQ</code>, <code>PL</code>, <code>PS</code>, and many more (including base‑mod Modification keys M... and their depths; see spec) :contentReference[oaicite:15]{index=15}.</p>
<h5 id="table-1-reserved-info-keys-excerpt">Table 1: Reserved INFO keys (excerpt)</h5>
<table>
<thead>
<tr>
<th>Key</th>
<th style="text-align: center;">Number</th>
<th style="text-align: left;">Type</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>AA</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">String</td>
<td style="text-align: left;">Ancestral allele</td>
</tr>
<tr>
<td>AC</td>
<td style="text-align: center;">A</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Allele count per ALT allele</td>
</tr>
<tr>
<td>AF</td>
<td style="text-align: center;">A</td>
<td style="text-align: left;">Float</td>
<td style="text-align: left;">Allele frequency per ALT allele</td>
</tr>
<tr>
<td>AN</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Total number of alleles called</td>
</tr>
<tr>
<td>DP</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Combined depth across samples</td>
</tr>
<tr>
<td>MQ</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Float</td>
<td style="text-align: left;">RMS mapping quality</td>
</tr>
<tr>
<td>...</td>
<td style="text-align: center;">...</td>
<td style="text-align: left;">...</td>
<td style="text-align: left;">...</td>
</tr>
</tbody>
</table>
<h5 id="table-2-reserved-formatgenotype-keys-excerpt">Table 2: Reserved FORMAT/Genotype keys (excerpt)</h5>
<table>
<thead>
<tr>
<th>Field</th>
<th style="text-align: center;">Number</th>
<th style="text-align: left;">Type</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>GT</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">String</td>
<td style="text-align: left;">Genotype (0=REF, 1...=ALTs; phasing <code>/</code> or <code>|</code>)</td>
</tr>
<tr>
<td>GQ</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Conditional genotype quality</td>
</tr>
<tr>
<td>DP</td>
<td style="text-align: center;">1</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Read depth</td>
</tr>
<tr>
<td>AD</td>
<td style="text-align: center;">R</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Read depth per allele</td>
</tr>
<tr>
<td>PL</td>
<td style="text-align: center;">G</td>
<td style="text-align: left;">Integer</td>
<td style="text-align: left;">Phred‐scaled genotype likelihoods</td>
</tr>
<tr>
<td>...</td>
<td style="text-align: center;">...</td>
<td style="text-align: left;">...</td>
<td style="text-align: left;">...</td>
</tr>
</tbody>
</table>
<p>For full lists and detailed definitions, see Sections 3–5 and Tables 1–2 in the spec.</p>
<h2 id="2-understanding-the-vcf-format-and-the-haplotype-representation">2 Understanding the VCF format and the haplotype representation</h2>
<p>VCF records use a simple haplotype representation to describe variant alleles at a locus.  A VCF record holds all segregating alleles—both reference and alternate—as well as, optionally, genotypes for multiple individuals at that locus.  ALT haplotypes are constructed from the REF haplotype by replacing the reference bases at <strong>POS</strong> with the alternate bases.  In essence, the VCF record specifies a-REF-t and the alternative haplotypes are a-ALT-t for each ALT allele.&#x20;</p>
<h3 id="21-vcf-tag-naming-conventions">2.1 VCF tag naming conventions</h3>
<p>Implementation‑defined tags should follow these suffix/prefix conventions:&#x20;</p>
<ul>
<li><strong><code>L</code> suffix</strong> — log₁₀‐likelihood (e.g., <code>GL</code>, <code>CNL</code>): log₁₀ Pr(Data|Model), negative numbers.</li>
<li><strong><code>P</code> suffix</strong> — linear‐scale posterior probability (e.g., <code>GP</code>, <code>CNP</code>): Pr(Model|Data).</li>
<li><strong><code>Q</code> suffix</strong> — phred‐scaled quality: −10 log₁₀ Pr(Data|Model) (e.g., <code>GQ</code>, <code>CNQ</code>); the fixed site‐level <strong>QUAL</strong> field uses the same scale.</li>
<li><strong><code>L</code> prefix</strong> — “local‐allele” equivalent of a Number=A, R, or G field (see 1.6.2).</li>
</ul>
<hr />
<h2 id="3-info-keys-used-for-structural-variants">3 INFO keys used for structural variants</h2>
<p>The following INFO keys are reserved for encoding structural variants.  Values for imprecise variants should be best estimates.  When per‑allele values are required, they must be provided for <strong>all</strong> ALT alleles (including non‑structural alleles), using the missing‐value placeholder (<code>.</code>) where appropriate.&#x20;</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>##INFO=&lt;ID=IMPRECISE,Number=0,Type=Flag,Description=&quot;Imprecise structural variation&quot;&gt;
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>##INFO=&lt;ID=NOVEL,Number=0,Type=Flag,Description=&quot;Indicates a novel structural variation&quot;&gt;
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>##INFO=&lt;ID=END,Number=1,Type=Integer,Description=&quot;Deprecated. Present for backwards compatibility&quot;&gt;
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>##INFO=&lt;ID=SVTYPE,Number=1,Type=String,Description=&quot;Type of structural variant&quot;&gt;
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>##INFO=&lt;ID=SVLEN,Number=A,Type=Integer,Description=&quot;Length of structural variant&quot;&gt;
<a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>##INFO=&lt;ID=CIPOS,Number=.,Type=Integer,Description=&quot;Confidence interval around POS&quot;&gt;
</code></pre></div>
<ul>
<li><strong>IMPRECISE</strong> (Flag): marks an imprecise structural‐variant ALT.  If a record is marked IMPRECISE, <code>CIPOS</code> (or <code>CIEND</code>) <strong>must</strong> be present (even if <code>0,0</code>).</li>
<li><strong>NOVEL</strong> (Flag): indicates a novel structural variant.</li>
<li><strong>END</strong> (Integer): deprecated in favor of <code>SVLEN</code> (and <code>LEN</code> in FORMAT).</li>
<li><strong>SVTYPE</strong> (String): deprecated (redundant with symbolic <code>ALT</code> alleles); see 1.4.5 for valid ALT symbols (e.g., <code>&lt;DEL&gt;</code>, <code>&lt;DUP&gt;</code>, <code>&lt;INV&gt;</code>).</li>
<li><strong>SVLEN</strong> (Integer, one per ALT): variant length.  For <code>&lt;DEL&gt;</code>, <code>&lt;INS&gt;</code>, <code>&lt;DUP&gt;</code>, <code>&lt;INV&gt;</code> alleles, the number of bases deleted/inserted/etc.; for <code>&lt;CNV&gt;</code>, the segment length.  Use <code>.</code> for other ALT alleles.</li>
<li><strong>CIPOS</strong> (Integer list): confidence interval around <strong>POS</strong>; twice as many entries as ALT alleles, giving start/end offsets for each.</li>
</ul>
<hr />
<h2 id="4-format-keys-used-for-structural-variants">4 FORMAT keys used for structural variants</h2>
<p>Reserved per‑sample FORMAT keys for structural variants include (among others):&#x20;</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>##FORMAT=&lt;ID=CN,Number=1,Type=Float,Description=&quot;Copy number&quot;&gt;
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>##FORMAT=&lt;ID=PSL,Number=.,Type=Integer,Description=&quot;Phase set list for structural variants&quot;&gt;
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>##FORMAT=&lt;ID=PSO,Number=.,Type=Integer,Description=&quot;Phase set offsets&quot;&gt;
</code></pre></div>
<ul>
<li><strong>CN</strong> (Float): estimated per‑sample copy number.</li>
<li><strong>PSL/PSO</strong> (Integer lists): phase sets and offsets, used to indicate phasing of SNVs with structural events (see 2.1).</li>
</ul>
<hr />
<h2 id="5-representing-variation-in-vcf-records">5 Representing variation in VCF records</h2>
<h3 id="51-creating-vcf-entries-for-snps-and-small-indels">5.1 Creating VCF entries for SNPs and small indels</h3>
<p>VCF entries for simple variants are constructed by aligning REF and ALT haplotypes:</p>
<h4 id="511-example-1">5.1.1 Example 1</h4>
<p><em>Single SNP at position 100:</em></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>chr1 100 . A G 50 PASS . GT 0/1
</code></pre></div>
<h4 id="512-example-2">5.1.2 Example 2</h4>
<p><em>Two‐base insertion at position 200:</em></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>chr1 200 . AT ATG 60 PASS . GT 1|0
</code></pre></div>
<h4 id="513-example-3">5.1.3 Example 3</h4>
<p><em>Two‐base deletion at position 300:</em></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>chr1 300 . AT A 70 PASS . GT 0/1
</code></pre></div>
<h3 id="52-decoding-vcf-entries-for-snps-and-small-indels">5.2 Decoding VCF entries for SNPs and small indels</h3>
<p>Given a REF/ALT and genotype, one reconstructs the haplotypes:</p>
<ul>
<li><strong>5.2.1 SNP</strong> — replace the single base at <strong>POS</strong>.</li>
<li><strong>5.2.2 Insertion</strong> — pad REF to first base, ALT contains inserted bases.</li>
<li><strong>5.2.3 Deletion</strong> — REF longer than ALT; ALT is the deleted‐bases string.</li>
<li><strong>5.2.4 Mixed (microsatellite)</strong> — multiple ALT alleles of varying lengths.</li>
</ul>
<h3 id="53-encoding-structural-variants">5.3 Encoding Structural Variants</h3>
<p>Symbolic alleles (e.g., <code>&lt;DEL&gt;</code>, <code>&lt;INS&gt;</code>, <code>&lt;DUP&gt;</code>) are used in <strong>ALT</strong>, with companion INFO/FORMAT fields (<code>SVTYPE</code>, <code>SVLEN</code>, <code>CIPOS</code>, <code>CN</code>, etc.) to capture imprecision, length, confidence intervals, copy‐number, and phasing information.</p>
<h3 id="54-specifying-complex-rearrangements-with-breakends">5.4 Specifying complex rearrangements with breakends</h3>
<p>Breakend notation allows representation of arbitrary rearrangements:</p>
<ul>
<li><strong>5.4.1 Inserted Sequence</strong> — specify sequence within <code>[]</code> brackets.</li>
<li><strong>5.4.2 Large Insertions</strong> — use symbolic <code>&lt;INS&gt;</code> + <code>SVLEN</code>.</li>
<li><strong>5.4.3 Multiple mates</strong> — list multiple breakend records for multi‐way events.</li>
<li><strong>5.4.4 Explicit partners</strong> — include partner chromosome/position within the ALT string.</li>
<li><strong>5.4.5 Telomeres</strong> — use <code>&lt;T&gt;</code> symbolic allele for telomeric breakends.</li>
<li><strong>5.4.6 Event modifiers</strong> — flags like <code>IMPRECISE</code>, <code>NOVEL</code>.</li>
<li><strong>5.4.7 Inversions</strong> — <code>&lt;INV&gt;</code> with <code>SVLEN</code>.</li>
<li><strong>5.4.8 Uncertainty around breakend location</strong> — use <code>CIPOS</code> and <code>CIEND</code>.</li>
<li><strong>5.4.9 Single breakends</strong> — breakend with no defined partner.</li>
</ul>
<p>For full details and all sub‑cases (e.g. nested inversions, sample mixtures), see Sections 5.4.1–5.4.9 in the spec.&#x20;</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.expand", "navigation.indexes", "navigation.top", "search.highlight", "search.share", "content.code.copy"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>