{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2c882ffc",
   "metadata": {},
   "source": [
    "# Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "73fd54b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from pygnome.parsers.msi.msi_sites_loader import load_msi_sites\n",
    "from pygnome.feature_store.genomic_feature_store import GenomicFeatureStore\n",
    "from pathlib import Path\n",
    "from pygnome.parsers.vcf import VcfReader\n",
    "from pygnome.parsers.vcf.ann import AnnParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e26f0dc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/kqrw311/workspace/pygnome/data/test.vcf')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA_DIR = Path('.') / 'data'\n",
    "\n",
    "vcf_file = DATA_DIR / 'test.vcf'\n",
    "vcf_file.absolute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30138887",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t10018\t.\tC\tCT\t85.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=Insertion;DP=47;VD=5;AF=0.1064;BIAS=2:2;REFBIAS=20:22;VARBIAS=2:3;PMEAN=16.6;PSTD=1;QUAL=36.8;QSTD=1;SBF=1;ODDRATIO=1.35477;MQ=37.8;SN=10;HIAF=0.125;ADJAF=0;SHIFT3=1;MSI=14;MSILEN=6;NM=1.2;HICNT=5;HICOV=40;LSEQ=NNNTAACCCTAACCCTAACC;RSEQ=TAACCCTAACCCTAACCCTA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=CT|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1851_-1850insT|||||1850|,CT|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4385_*4386insA|||||4385|,CT|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10018_10019insT||||||;EPR=pass;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:47:5:42,5:0.1064:20,22:2,3\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: Insertion\n",
      "\tDP: 47\n",
      "\tVD: 5\n",
      "\tAF: 0.1064\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 20:22\n",
      "\tVARBIAS: 2:3\n",
      "\tPMEAN: 16.6\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 36.8\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 1.0\n",
      "\tODDRATIO: 1.35477\n",
      "\tMQ: 37.8\n",
      "\tSN: 10.0\n",
      "\tHIAF: 0.125\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 1\n",
      "\tMSI: 14.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 1.2\n",
      "\tHICNT: 5\n",
      "\tHICOV: 40\n",
      "\tLSEQ: NNNTAACCCTAACCCTAACC\n",
      "\tRSEQ: TAACCCTAACCCTAACCCTA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['CT|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1851_-1850insT|||||1850|', 'CT|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4385_*4386insA|||||4385|', 'CT|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10018_10019insT||||||']\n",
      "\tEPR: ['pass']\n",
      "\tANNOTATED: True\n",
      "\tInsertion(variant_chr1_10017, chr1:10017-10018, C>CT)\n",
      "\tAllele: CT | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1851_-1850insT\n",
      "\tAllele: CT | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4385_*4386insA\n",
      "\tAllele: CT | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10018_10019insT\n",
      "chr1\t10057\t.\tA\tC\t134.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=133;VD=15;AF=0.1128;BIAS=2:2;REFBIAS=55:58;VARBIAS=6:9;PMEAN=33.1;PSTD=1;QUAL=34.4;QSTD=1;SBF=0.59094;ODDRATIO=1.41854;MQ=51.1;SN=4;HIAF=0.0984;ADJAF=0;SHIFT3=0;MSI=13;MSILEN=6;NM=3.2;HICNT=12;HICOV=122;LSEQ=TAACCCTAACCCTAACCCTA;RSEQ=CCCTAACCCTAACCCTAACC;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1812A>C|||||1812|,C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4347T>G|||||4347|,C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10057A>C||||||;EPR=pass;DBSNP_RS=1570391741;DBSNP_COMMON;GNOMAD_grpmax_joint=nfe;GNOMAD_AF_grpmax_joint=3.77986E-5;GNOMAD_AF_joint=2.64233E-5;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:133:15:113,15:0.1128:55,58:6,9\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 133\n",
      "\tVD: 15\n",
      "\tAF: 0.1128\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 55:58\n",
      "\tVARBIAS: 6:9\n",
      "\tPMEAN: 33.1\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 34.4\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.59094\n",
      "\tODDRATIO: 1.41854\n",
      "\tMQ: 51.1\n",
      "\tSN: 4.0\n",
      "\tHIAF: 0.0984\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 0\n",
      "\tMSI: 13.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 3.2\n",
      "\tHICNT: 12\n",
      "\tHICOV: 122\n",
      "\tLSEQ: TAACCCTAACCCTAACCCTA\n",
      "\tRSEQ: CCCTAACCCTAACCCTAACC\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1812A>C|||||1812|', 'C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4347T>G|||||4347|', 'C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10057A>C||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 1570391741\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: nfe\n",
      "\tGNOMAD_AF_grpmax_joint: 3.77986e-05\n",
      "\tGNOMAD_AF_joint: 2.64233e-05\n",
      "\tANNOTATED: True\n",
      "\tSNP(variant_chr1_10056, chr1:10056-10057, A>C)\n",
      "\tAllele: C | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1812A>C\n",
      "\tAllele: C | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4347T>G\n",
      "\tAllele: C | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10057A>C\n",
      "chr1\t10146\trs779258992\tAC\tA\t152.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=Deletion;DP=167;VD=16;AF=0.0958;BIAS=2:2;REFBIAS=62:67;VARBIAS=6:10;PMEAN=26.2;PSTD=1;QUAL=38;QSTD=1;SBF=0.59659;ODDRATIO=1.53776;MQ=32.3;SN=32;HIAF=0.1088;ADJAF=0;SHIFT3=3;MSI=11;MSILEN=6;NM=2.4;HICNT=16;HICOV=147;LSEQ=TAACCCTAACCCTAACCCTA;RSEQ=CCCTAACCCTAACCCTAACC;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1722delC|||||1722|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4257delG|||||4257|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10147delC||||||;EPR=pass;DBSNP_RS=779258992;DBSNP_COMMON;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:167:16:129,16:0.0958:62,67:6,10\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: Deletion\n",
      "\tDP: 167\n",
      "\tVD: 16\n",
      "\tAF: 0.0958\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 62:67\n",
      "\tVARBIAS: 6:10\n",
      "\tPMEAN: 26.2\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 38.0\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.59659\n",
      "\tODDRATIO: 1.53776\n",
      "\tMQ: 32.3\n",
      "\tSN: 32.0\n",
      "\tHIAF: 0.1088\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 3\n",
      "\tMSI: 11.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 2.4\n",
      "\tHICNT: 16\n",
      "\tHICOV: 147\n",
      "\tLSEQ: TAACCCTAACCCTAACCCTA\n",
      "\tRSEQ: CCCTAACCCTAACCCTAACC\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1722delC|||||1722|', 'A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4257delG|||||4257|', 'A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10147delC||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 779258992\n",
      "\tDBSNP_COMMON: True\n",
      "\tANNOTATED: True\n",
      "\tDeletion(rs779258992, chr1:10145-10147, AC>A)\n",
      "\tAllele: A | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1722delC\n",
      "\tAllele: A | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4257delG\n",
      "\tAllele: A | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10147delC\n",
      "chr1\t10150\trs371194064\tC\tT\t154.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=153;VD=20;AF=0.1307;BIAS=2:2;REFBIAS=69:61;VARBIAS=10:10;PMEAN=31;PSTD=1;QUAL=35.8;QSTD=1;SBF=0.81467;ODDRATIO=1.13022;MQ=35.2;SN=5.667;HIAF=0.125;ADJAF=0;SHIFT3=0;MSI=1;MSILEN=1;NM=2;HICNT=17;HICOV=136;LSEQ=CCTAACCCTAACCCTAACCC;RSEQ=TAACCCTAACCCTAACCCTA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=T|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1719C>T|||||1719|,T|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4254G>A|||||4254|,T|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10150C>T||||||;EPR=pass;DBSNP_RS=371194064;DBSNP_COMMON;GNOMAD_grpmax_joint=eas;GNOMAD_AF_grpmax_joint=0.0185185;GNOMAD_AF_joint=0.00796178;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:153:20:130,20:0.1307:69,61:10,10\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 153\n",
      "\tVD: 20\n",
      "\tAF: 0.1307\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 69:61\n",
      "\tVARBIAS: 10:10\n",
      "\tPMEAN: 31.0\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 35.8\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.81467\n",
      "\tODDRATIO: 1.13022\n",
      "\tMQ: 35.2\n",
      "\tSN: 5.667\n",
      "\tHIAF: 0.125\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 0\n",
      "\tMSI: 1.0\n",
      "\tMSILEN: 1.0\n",
      "\tNM: 2.0\n",
      "\tHICNT: 17\n",
      "\tHICOV: 136\n",
      "\tLSEQ: CCTAACCCTAACCCTAACCC\n",
      "\tRSEQ: TAACCCTAACCCTAACCCTA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['T|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1719C>T|||||1719|', 'T|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4254G>A|||||4254|', 'T|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10150C>T||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 371194064\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: eas\n",
      "\tGNOMAD_AF_grpmax_joint: 0.0185185\n",
      "\tGNOMAD_AF_joint: 0.00796178\n",
      "\tANNOTATED: True\n",
      "\tSNP(rs371194064, chr1:10149-10150, C>T)\n",
      "\tAllele: T | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1719C>T\n",
      "\tAllele: T | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4254G>A\n",
      "\tAllele: T | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10150C>T\n",
      "chr1\t10222\trs1425505601\tT\tC\t152.0\tNM5.25\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=173;VD=16;AF=0.0925;BIAS=2:2;REFBIAS=68:74;VARBIAS=3:13;PMEAN=39.6;PSTD=1;QUAL=38.2;QSTD=1;SBF=0.03334;ODDRATIO=3.95193;MQ=50.9;SN=15;HIAF=0.0938;ADJAF=0;SHIFT3=0;MSI=6;MSILEN=6;NM=6;HICNT=15;HICOV=160;LSEQ=CCTAACCCTAACCCTAACCC;RSEQ=AACCCTAACCCCTAACCCTA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1647T>C|||||1647|,C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4182A>G|||||4182|,C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10222T>C||||||;EPR=pass;DBSNP_RS=1425505601;DBSNP_COMMON;GNOMAD_grpmax_joint=nfe;GNOMAD_AF_grpmax_joint=1.30668E-4;GNOMAD_AF_joint=1.15323E-4;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:173:16:142,16:0.0925:68,74:3,13\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 173\n",
      "\tVD: 16\n",
      "\tAF: 0.0925\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 68:74\n",
      "\tVARBIAS: 3:13\n",
      "\tPMEAN: 39.6\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 38.2\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.03334\n",
      "\tODDRATIO: 3.95193\n",
      "\tMQ: 50.9\n",
      "\tSN: 15.0\n",
      "\tHIAF: 0.0938\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 0\n",
      "\tMSI: 6.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 6.0\n",
      "\tHICNT: 15\n",
      "\tHICOV: 160\n",
      "\tLSEQ: CCTAACCCTAACCCTAACCC\n",
      "\tRSEQ: AACCCTAACCCCTAACCCTA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1647T>C|||||1647|', 'C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4182A>G|||||4182|', 'C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10222T>C||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 1425505601\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: nfe\n",
      "\tGNOMAD_AF_grpmax_joint: 0.000130668\n",
      "\tGNOMAD_AF_joint: 0.000115323\n",
      "\tANNOTATED: True\n",
      "\tSNP(rs1425505601, chr1:10221-10222, T>C)\n",
      "\tAllele: C | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1647T>C\n",
      "\tAllele: C | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4182A>G\n",
      "\tAllele: C | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10222T>C\n",
      "chr1\t10228\trs1429280313\tT\tC\t125.0\tNM5.25\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=169;VD=12;AF=0.071;BIAS=2:2;REFBIAS=63:80;VARBIAS=2:10;PMEAN=44.8;PSTD=1;QUAL=35;QSTD=1;SBF=0.07492;ODDRATIO=3.90829;MQ=52;SN=11;HIAF=0.0743;ADJAF=0;SHIFT3=0;MSI=5;MSILEN=6;NM=5.8;HICNT=11;HICOV=148;LSEQ=CCTAACCCTAACCCTAACCC;RSEQ=AACCCCTAACCCTAACCCTA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1641T>C|||||1641|,C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4176A>G|||||4176|,C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10228T>C||||||;EPR=pass;DBSNP_RS=1429280313;DBSNP_COMMON;GNOMAD_grpmax_joint=nfe;GNOMAD_AF_grpmax_joint=6.48298E-4;GNOMAD_AF_joint=4.90773E-4;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:169:12:143,12:0.071:63,80:2,10\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 169\n",
      "\tVD: 12\n",
      "\tAF: 0.071\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 63:80\n",
      "\tVARBIAS: 2:10\n",
      "\tPMEAN: 44.8\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 35.0\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.07492\n",
      "\tODDRATIO: 3.90829\n",
      "\tMQ: 52.0\n",
      "\tSN: 11.0\n",
      "\tHIAF: 0.0743\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 0\n",
      "\tMSI: 5.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 5.8\n",
      "\tHICNT: 11\n",
      "\tHICOV: 148\n",
      "\tLSEQ: CCTAACCCTAACCCTAACCC\n",
      "\tRSEQ: AACCCCTAACCCTAACCCTA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1641T>C|||||1641|', 'C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4176A>G|||||4176|', 'C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10228T>C||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 1429280313\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: nfe\n",
      "\tGNOMAD_AF_grpmax_joint: 0.000648298\n",
      "\tGNOMAD_AF_joint: 0.000490773\n",
      "\tANNOTATED: True\n",
      "\tSNP(rs1429280313, chr1:10227-10228, T>C)\n",
      "\tAllele: C | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1641T>C\n",
      "\tAllele: C | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4176A>G\n",
      "\tAllele: C | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10228T>C\n",
      "chr1\t10230\trs775928745\tAC\tA\t178.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=Deletion;DP=171;VD=22;AF=0.1287;BIAS=2:2;REFBIAS=43:63;VARBIAS=11:11;PMEAN=39.1;PSTD=1;QUAL=40;QSTD=1;SBF=0.48007;ODDRATIO=1.46064;MQ=28;SN=44;HIAF=0.156;ADJAF=0.0292;SHIFT3=3;MSI=11;MSILEN=6;NM=2.4;HICNT=22;HICOV=141;LSEQ=TAACCCTAACCCTAACCCTA;RSEQ=CCCTAACCCTAACCCTAAAC;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1638delC|||||1638|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4173delG|||||4173|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10231delC||||||;EPR=pass;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:171:22:106,22:0.1287:43,63:11,11\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: Deletion\n",
      "\tDP: 171\n",
      "\tVD: 22\n",
      "\tAF: 0.1287\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 43:63\n",
      "\tVARBIAS: 11:11\n",
      "\tPMEAN: 39.1\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 40.0\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.48007\n",
      "\tODDRATIO: 1.46064\n",
      "\tMQ: 28.0\n",
      "\tSN: 44.0\n",
      "\tHIAF: 0.156\n",
      "\tADJAF: 0.0292\n",
      "\tSHIFT3: 3\n",
      "\tMSI: 11.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 2.4\n",
      "\tHICNT: 22\n",
      "\tHICOV: 141\n",
      "\tLSEQ: TAACCCTAACCCTAACCCTA\n",
      "\tRSEQ: CCCTAACCCTAACCCTAAAC\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1638delC|||||1638|', 'A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4173delG|||||4173|', 'A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10231delC||||||']\n",
      "\tEPR: ['pass']\n",
      "\tANNOTATED: True\n",
      "\tDeletion(rs775928745, chr1:10229-10231, AC>A)\n",
      "\tAllele: A | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1638delC\n",
      "\tAllele: A | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4173delG\n",
      "\tAllele: A | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10231delC\n",
      "chr1\t10231\trs200279319\tC\tA\t154.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=171;VD=24;AF=0.1404;BIAS=2:2;REFBIAS=43:63;VARBIAS=11:13;PMEAN=32.4;PSTD=1;QUAL=33.8;QSTD=1;SBF=0.65299;ODDRATIO=1.23764;MQ=37.8;SN=7;HIAF=0.1489;ADJAF=0;SHIFT3=2;MSI=5;MSILEN=6;NM=3.5;HICNT=21;HICOV=141;LSEQ=AACCCTAACCCTAACCCTAA;RSEQ=CCCTAACCCTAACCCTAAAC;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1638C>A|||||1638|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4173G>T|||||4173|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10231C>A||||||;EPR=pass;DBSNP_RS=200279319;DBSNP_COMMON;GNOMAD_grpmax_joint=amr;GNOMAD_AF_grpmax_joint=0.166667;GNOMAD_AF_joint=0.0357143;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:171:24:106,24:0.1404:43,63:11,13\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 171\n",
      "\tVD: 24\n",
      "\tAF: 0.1404\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 43:63\n",
      "\tVARBIAS: 11:13\n",
      "\tPMEAN: 32.4\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 33.8\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 0.65299\n",
      "\tODDRATIO: 1.23764\n",
      "\tMQ: 37.8\n",
      "\tSN: 7.0\n",
      "\tHIAF: 0.1489\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 2\n",
      "\tMSI: 5.0\n",
      "\tMSILEN: 6.0\n",
      "\tNM: 3.5\n",
      "\tHICNT: 21\n",
      "\tHICOV: 141\n",
      "\tLSEQ: AACCCTAACCCTAACCCTAA\n",
      "\tRSEQ: CCCTAACCCTAACCCTAAAC\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1638C>A|||||1638|', 'A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4173G>T|||||4173|', 'A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10231C>A||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 200279319\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: amr\n",
      "\tGNOMAD_AF_grpmax_joint: 0.166667\n",
      "\tGNOMAD_AF_joint: 0.0357143\n",
      "\tANNOTATED: True\n",
      "\tSNP(rs200279319, chr1:10230-10231, C>A)\n",
      "\tAllele: A | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1638C>A\n",
      "\tAllele: A | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4173G>T\n",
      "\tAllele: A | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10231C>A\n",
      "chr1\t10235\trs1035249121\tT\tA\t124.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=172;VD=16;AF=0.093;BIAS=2:2;REFBIAS=54:68;VARBIAS=7:9;PMEAN=40.4;PSTD=1;QUAL=31;QSTD=1;SBF=1;ODDRATIO=1.02086;MQ=42.5;SN=3;HIAF=0.0909;ADJAF=0;SHIFT3=1;MSI=2;MSILEN=1;NM=3;HICNT=12;HICOV=132;LSEQ=CTAACCCTAACCCTAACCCC;RSEQ=AACCCTAACCCTAAACCCTA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1634T>A|||||1634|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4169A>T|||||4169|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10235T>A||||||;EPR=pass;DBSNP_RS=1035249121;DBSNP_COMMON;GNOMAD_grpmax_joint=nfe;GNOMAD_AF_grpmax_joint=0.00847458;GNOMAD_AF_joint=0.00555556;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:172:16:122,16:0.093:54,68:7,9\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 172\n",
      "\tVD: 16\n",
      "\tAF: 0.093\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 54:68\n",
      "\tVARBIAS: 7:9\n",
      "\tPMEAN: 40.4\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 31.0\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 1.0\n",
      "\tODDRATIO: 1.02086\n",
      "\tMQ: 42.5\n",
      "\tSN: 3.0\n",
      "\tHIAF: 0.0909\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 1\n",
      "\tMSI: 2.0\n",
      "\tMSILEN: 1.0\n",
      "\tNM: 3.0\n",
      "\tHICNT: 12\n",
      "\tHICOV: 132\n",
      "\tLSEQ: CTAACCCTAACCCTAACCCC\n",
      "\tRSEQ: AACCCTAACCCTAAACCCTA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1634T>A|||||1634|', 'A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4169A>T|||||4169|', 'A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10235T>A||||||']\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 1035249121\n",
      "\tDBSNP_COMMON: True\n",
      "\tGNOMAD_grpmax_joint: nfe\n",
      "\tGNOMAD_AF_grpmax_joint: 0.00847458\n",
      "\tGNOMAD_AF_joint: 0.00555556\n",
      "\tANNOTATED: True\n",
      "\tSNP(rs1035249121, chr1:10234-10235, T>A)\n",
      "\tAllele: A | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1634T>A\n",
      "\tAllele: A | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4169A>T\n",
      "\tAllele: A | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10235T>A\n",
      "chr1\t10236\t.\tAACCCT\tA\t166.0\tPASS\tSAMPLE=LU2551_G01-38211;TYPE=Deletion;DP=172;VD=18;AF=0.1047;BIAS=2:2;REFBIAS=51:69;VARBIAS=7:11;PMEAN=29.1;PSTD=1;QUAL=40;QSTD=0;SBF=0.80461;ODDRATIO=1.16024;MQ=44.9;SN=36;HIAF=0.1286;ADJAF=0;SHIFT3=1;MSI=3;MSILEN=1;NM=1.4;HICNT=18;HICOV=140;LSEQ=TAACCCTAACCCTAACCCCT;RSEQ=AACCCTAAACCCTAAACCCT;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1632_-1628delACCCT|||||1632|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4163_*4167delAGGGT|||||4167|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10237_10241delACCCT||||||;EPR=pass;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:172:18:120,18:0.1047:51,69:7,11\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: Deletion\n",
      "\tDP: 172\n",
      "\tVD: 18\n",
      "\tAF: 0.1047\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 51:69\n",
      "\tVARBIAS: 7:11\n",
      "\tPMEAN: 29.1\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 40.0\n",
      "\tQSTD: 0.0\n",
      "\tSBF: 0.80461\n",
      "\tODDRATIO: 1.16024\n",
      "\tMQ: 44.9\n",
      "\tSN: 36.0\n",
      "\tHIAF: 0.1286\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 1\n",
      "\tMSI: 3.0\n",
      "\tMSILEN: 1.0\n",
      "\tNM: 1.4\n",
      "\tHICNT: 18\n",
      "\tHICOV: 140\n",
      "\tLSEQ: TAACCCTAACCCTAACCCCT\n",
      "\tRSEQ: AACCCTAAACCCTAAACCCT\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1632_-1628delACCCT|||||1632|', 'A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4163_*4167delAGGGT|||||4167|', 'A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10237_10241delACCCT||||||']\n",
      "\tEPR: ['pass']\n",
      "\tANNOTATED: True\n",
      "\tDeletion(variant_chr1_10235, chr1:10235-10241, AACCCT>A)\n",
      "\tAllele: A | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1632_-1628delACCCT\n",
      "\tAllele: A | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4163_*4167delAGGGT\n",
      "\tAllele: A | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10237_10241delACCCT\n",
      "chr1\t10237\tCOSV70831099;rs1490526935\tA\tC\t109.0\tNM5.25\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=172;VD=9;AF=0.0523;BIAS=2:2;REFBIAS=51:69;VARBIAS=4:5;PMEAN=32.4;PSTD=1;QUAL=34.7;QSTD=1;SBF=1;ODDRATIO=1.08169;MQ=37.1;SN=18;HIAF=0.0643;ADJAF=0;SHIFT3=2;MSI=3;MSILEN=1;NM=6.2;HICNT=9;HICOV=140;LSEQ=AACCCTAACCCTAACCCCTA;RSEQ=CCCTAACCCTAAACCCTAAA;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1632A>C|||||1632|,C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4167T>G|||||4167|,C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10237A>C||||||;CNT=1;GENE=WASH7P;EPR=pass;DBSNP_RS=1490526935;DBSNP_COMMON;COSMIC_LEGACY_ID=COSN8882887;COSMIC_CNT=1;COSMIC_ID=COSV70831099;GNOMAD_grpmax_joint=nfe;GNOMAD_AF_grpmax_joint=4.42087E-4;GNOMAD_AF_joint=2.63296E-4;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:172:9:120,9:0.0523:51,69:4,5\n",
      "\tSAMPLE: LU2551_G01-38211\n",
      "\tTYPE: SNV\n",
      "\tDP: 172\n",
      "\tVD: 9\n",
      "\tAF: 0.0523\n",
      "\tBIAS: 2:2\n",
      "\tREFBIAS: 51:69\n",
      "\tVARBIAS: 4:5\n",
      "\tPMEAN: 32.4\n",
      "\tPSTD: 1.0\n",
      "\tQUAL: 34.7\n",
      "\tQSTD: 1.0\n",
      "\tSBF: 1.0\n",
      "\tODDRATIO: 1.08169\n",
      "\tMQ: 37.1\n",
      "\tSN: 18.0\n",
      "\tHIAF: 0.0643\n",
      "\tADJAF: 0.0\n",
      "\tSHIFT3: 2\n",
      "\tMSI: 3.0\n",
      "\tMSILEN: 1.0\n",
      "\tNM: 6.2\n",
      "\tHICNT: 9\n",
      "\tHICOV: 140\n",
      "\tLSEQ: AACCCTAACCCTAACCCCTA\n",
      "\tRSEQ: CCCTAACCCTAAACCCTAAA\n",
      "\tDUPRATE: 0.0\n",
      "\tSPLITREAD: 0\n",
      "\tSPANPAIR: 0\n",
      "\tANN: ['C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1632A>C|||||1632|', 'C|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4167T>G|||||4167|', 'C|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10237A>C||||||']\n",
      "\tCNT: 1\n",
      "\tGENE: WASH7P\n",
      "\tEPR: ['pass']\n",
      "\tDBSNP_RS: 1490526935\n",
      "\tDBSNP_COMMON: True\n",
      "\tCOSMIC_LEGACY_ID: COSN8882887\n",
      "\tCOSMIC_CNT: 1\n",
      "\tCOSMIC_ID: COSV70831099\n",
      "\tGNOMAD_grpmax_joint: nfe\n",
      "\tGNOMAD_AF_grpmax_joint: 0.000442087\n",
      "\tGNOMAD_AF_joint: 0.000263296\n",
      "\tANNOTATED: True\n",
      "\tSNP(COSV70831099;rs1490526935, chr1:10236-10237, A>C)\n",
      "\tAllele: C | Effect: upstream_gene_variant | Impact: MODIFIER | Gene: DDX11L1 | transcript: ENST00000456328.2 | HGVS.c: n.-1632A>C\n",
      "\tAllele: C | Effect: downstream_gene_variant | Impact: MODIFIER | Gene: WASH7P | transcript: ENST00000488147.1 | HGVS.c: n.*4167T>G\n",
      "\tAllele: C | Effect: intergenic_region | Impact: MODIFIER | Gene: CHR_START-DDX11L1 | custom: CHR_START-ENSG00000223972 | HGVS.c: n.10237A>C\n",
      "chr1\t10241\t.\tT\tA\t126.0\tNM5.25\tSAMPLE=LU2551_G01-38211;TYPE=SNV;DP=170;VD=9;AF=0.0529;BIAS=2:2;REFBIAS=44:55;VARBIAS=1:8;PMEAN=32.7;PSTD=1;QUAL=40;QSTD=0;SBF=0.07696;ODDRATIO=6.31665;MQ=55.6;SN=18;HIAF=0.0857;ADJAF=0;SHIFT3=1;MSI=2;MSILEN=1;NM=5.9;HICNT=9;HICOV=105;LSEQ=CTAACCCTAACCCCTAACCC;RSEQ=AACCCTAAACCCTAAACCCT;DUPRATE=0;SPLITREAD=0;SPANPAIR=0;ANN=A|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.-1628T>A|||||1628|,A|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*4163A>T|||||4163|,A|intergenic_region|MODIFIER|CHR_START-DDX11L1|CHR_START-ENSG00000223972|intergenic_region|CHR_START-ENSG00000223972|||n.10241T>A||||||;EPR=pass;DBSNP_RS=960927773;DBSNP_COMMON;GNOMAD_grpmax_joint=afr;GNOMAD_AF_grpmax_joint=0.03125;GNOMAD_AF_joint=0.00925926;ANNOTATED\tGT:DP:VD:AD:AF:RD:ALD\t0/1:170:9:99,9:0.0529:44,55:1,8\n"
     ]
    }
   ],
   "source": [
    "for i, vr in enumerate(VcfReader(vcf_file)):\n",
    "    print(vr)\n",
    "    if i > 10:\n",
    "        break\n",
    "\n",
    "    for k, v in vr.info:\n",
    "        print(f\"\\t{k}: {vr.info[k]}\")\n",
    "\n",
    "    for v in vr:\n",
    "        print(f\"\\t{v}\")\n",
    "\n",
    "    for vann in vr.variant_annotations():\n",
    "        print(f\"\\t{vann}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc61856b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
